theano/tensor/nnet/abstract_conv.py

```diff
@@ -1540,6 +1587,39 @@ def conv(self, img, kern, mode="valid", dilation=1):
             raise NotImplementedError('only 2D and 3D convolution are implemented')
         return out
 
+    def unshared2d(self, inp, kern, out_shape, direction="forward"):
+        '''
+        Basic slow Python unshared 2d convolution.
+        '''
+        if self.convdim != 2:
+            raise NotImplementedError('Unshared convolution not implemented for %dD'
+                                      % self.convdim)
+        if direction == "forward":
+            if (kern.shape[0], kern.shape[1]) != (out_shape[0], out_shape[1]):
+                raise ValueError('Kernel shape ({},{}) does not match '
+                                 'output size ({},{})'.format(kern.shape[0], kern.shape[1],
+                                                              out_shape[0], out_shape[1]))
+        out = np.zeros(out_shape, dtype=inp.dtype)
+
+        for row in xrange(out_shape[0]):
+            for col in xrange(out_shape[1]):
+                if direction == "forward":
+                    out[row, col] += np.sum(np.multiply(inp[row:row + kern.shape[2],
+                                                        col:col + kern.shape[3]],
+                                            kern[row, col, ::-1, ::-1]))
+                elif direction == "backprop weights":
+                    out[row, col, ...] += kern[row, col] * \
+                        inp[row:row + out_shape[2], col:col + out_shape[3]]
+                elif direction == "backprop inputs":
+                    if row < kern.shape[0] and col < kern.shape[1]:
```

Comment:
I appreciate your effort not to duplicate code, but I still think this solution is less than ideal. It feels a bit silly to have two loops that go on for too long, and then correct this with an if that checks the indices.

There may also be a conceptual problem: your for loops suggest that you're iterating over the rows and columns of `out`, while in fact for the grad wrt inputs you're iterating over the rows and columns of `kern` (which just happens to be slightly smaller than `out`).

So, I would propose to move the `for row` and `for col` loops inside the ifs, so that they're specific to each direction. This adds three lines of code, but is easier to understand and does not have loops that are longer than necessary. (As an added bonus it tests the value of `direction` only once.)